search2/python/search/utils/beanstalkd.py

```diff
@@ -0,0 +1,155 @@
+"""TODO: Add description here"""
+from glob import glob
+from os import path
+import time
+import traceback
+from typing import Any, NamedTuple, List, Callable, Optional
+
+import asyncio
+from orjson import loads, dumps  # pylint: disable=no-name-in-module
+from pystalk import BeanstalkClient, BeanstalkError  # type: ignore
+from opensearchpy.exceptions import NotFoundError
+
+class Message(NamedTuple):
+    """TODO: Add description here"""
+    event: str
+    queueID: str
+    submissionID: str
+    data: Optional[Any]
+
+class Publisher(NamedTuple):
+    """TODO: Add description here"""
+
+    host: str
+    port: str
+    queue: str
+    message: Message
+
+class QueueConf(NamedTuple):
+    """TODO: Add description here"""
+
+    address: List[str]
+    events: dict
+    tubes: dict
+
+    def split_host_port(self):
+        """TODO: Add description here"""
+        hosts = []
+        ports = []
+        for host in self.address:
+            host, port = host.split(":")
+            hosts.append(host)
+            ports.append(port)
+        return hosts, ports
+
+def get_config_file_path(config_path_base_dir: str, assembly: str, suffix: str = ".y*ml"):
+    """TODO: Add description here"""
+    paths = glob(path.join(config_path_base_dir, assembly + suffix))
+
+    if not paths:
+        raise ValueError(
+            f"\n\nNo config path found for the assembly {assembly}. Exiting\n\n"
+        )
+
+    if len(paths) > 1:
+        print("\n\nMore than 1 config path found, choosing first")
+
+    return paths[0]
+
+def _specify_publisher(
+    submission_id, job_id, publisher_host, publisher_port, progress_event, event_queue
+):
+    return Publisher(
+            host = publisher_host,
+            port = publisher_port,
+            queue = event_queue,
+            message = Message(
+                event = progress_event,
+                queueID = job_id,
+                submissionID = submission_id,
+                data = None,
+            ))
+
+def listen(
+    handler_fn: Callable,
+    submit_msg_fn: Callable,
+    completed_msg_fn: Callable,
+    queue_conf: QueueConf,
+    tube: str,
+):
+    """TODO: Listen on beanstalkd here directly"""
+    hosts, ports = queue_conf.split_host_port()
+
+    for event in ("progress", "failed", "started", "completed"):
+        assert event in queue_conf.events
+
+    tube_conf = queue_conf.tubes[tube]
+    events_conf = queue_conf.events
+    clients = tuple(
```

Comment:
why is this a tuple?

Reply:
because this is an immutable collection

Reply:
gotcha, asking because of this distinction: https://stackoverflow.com/questions/24854139/lists-are-for-homogeneous-data-and-tuples-are-for-heterogeneous-data-why

Not asking for changes to the code but maybe worth a comment